core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -1217,6 +1260,71 @@ private[spark] class TaskSetManager(
   def executorAdded(): Unit = {
     recomputeLocality()
   }
+
+  /**
+   * A class for checking inefficient tasks to be speculated, the inefficient tasks come from
+   * the tasks which may be speculated by the previous strategy.
+   */
+  private[scheduler] class InefficientTaskCalculator {
+    private var allTotalRecordsRead = 0L
+    private var allTotalExecutorRunTime = 0L
+    @volatile private var successTaskProgressThreshold = 0.0D
+    private val runingTasksProgressRate = new ConcurrentHashMap[Long, Double]()
+
+    private[scheduler] def updateTaskProgressThreshold(result: DirectTaskResult[_]): Unit = {
+      var totalRecordsRead = 0L
+      var totalExecutorRunTime = 0L
+      result.accumUpdates.foreach { a =>
+        if (a.name == Some(shuffleRead.RECORDS_READ) ||
+          a.name == Some(input.RECORDS_READ)) {
+          val acc = a.asInstanceOf[LongAccumulator]
+          totalRecordsRead += acc.value
+        } else if (a.name == Some(InternalAccumulator.EXECUTOR_RUN_TIME)) {
+          val acc = a.asInstanceOf[LongAccumulator]
+          totalExecutorRunTime = acc.value
+        }
+      }
+      allTotalRecordsRead += totalRecordsRead
+      allTotalExecutorRunTime += totalExecutorRunTime
+      if (allTotalRecordsRead > 0 && allTotalExecutorRunTime > 0) {
+        successTaskProgressThreshold = allTotalRecordsRead / (allTotalExecutorRunTime / 1000.0)
+      }
+    }
+
+    private[scheduler] def updateRuningTasksProgressRate(
+        taskId: Long,
+        taskProgressRate: Double): Unit = {
+      runingTasksProgressRate.put(taskId, taskProgressRate)
+    }
+
+    private[scheduler] def removeRuningTasksProgressRate(taskId: Long): Unit = {
+      runingTasksProgressRate.remove(taskId)
+    }
+
+    private[scheduler] def maySpeculateTask(
+        tid: Long,
+        runtimeMs: Long,
+        taskInfo: TaskInfo): Boolean = {
+      // Only check inefficient tasks when successTaskProgressThreshold > 0, because some stage
+      // tasks may have neither input records nor shuffleRead records, so the
+      // successTaskProgressThreshold may be zero all the time, this case we should make sure
+      // it can be speculated. eg: some spark-sql like that 'msck repair table' or 'drop table'
+      // and so on.
+      lazy val currentTaskProgressRate = runingTasksProgressRate.getOrDefault(tid, 0.0)
+      if (successTaskProgressThreshold <= 0.0 || currentTaskProgressRate <= 0.0) {
```

Comment:
nit: Reorganize code so that we remove use of `lazy val` ?
Except for the case of `successTaskProgressThreshold <= 0`, we always use this immediately.
```suggestion
      if (successTaskProgressThreshold <= 0.0) return true
      
      val currentTaskProgressRate = runingTasksProgressRate.getOrDefault(tid, 0.0)
      if (currentTaskProgressRate <= 0.0) {
```